Assigning select fix

VFLashM opened the pull request:
Please take a look at this pull request.
Currently any assigning select (or `set`) breaks result set.
Assigning select sends no result token, but does send TDS_DONE_TOKEN with both TDS_DONE_MORE_RESULTS and TDS_DONE_COUNT flags.
As a result `find_result_or_done` exits prematurely and the rest of result sets are ignored.
Change I made fixes the problem, but I'm not sure about what to do with other functions like `process_rpc`.

denisenkom commented:
I think this will break UPDATE INSERT and DELETE queries, those queries return valid row count.
What does updating select return in rowcount?
I think correct way to handle this situation is to call nextset, no?

VFLashM commented:
Updating select returns rowcount of 1 (but does not return any rows).
I'd prefer to avoid calling nextset for these cases, but that's a solution too.

Before my change if was returning False from `find_result_or_done` marking it as end of stream.
After that it was impossible to get any other results.

You're right, it seems that this change have broken something.
I'll find it and add a test for it soon.

VFLashM commented:
Nope, sorry, the error I mentioned was unrelated to this change.

Regarding the change. I'm surely could have missed something, but all tests are passing right now.

Before logic was:
if received done token, stop all processing and close the stream (i.e. return false) unless there is more data _and_ current dataset returned no rows

Meaning it skips everything even if more data is presented when current data has rowcount.

It doesn't sound right. nextset should obviously return true if we have more data.

VFLashM commented:
For example, before, for given query logic was:

````
select 1 as a into #test
select * from #test
update #test set a = 2
select * from #test
````

1. we call `execute`, which calls `find_result_or_done`
3. `find_result_or_done` finds DONE token for first line: `select 1 as a into #test` and returns False, `execute` ignores result
3. because `find_result_or_done` exited before it found RESULT token, we have no description and no resultset
4. we call `nextset`, which calls `find_result_or_done`, it finds RESULT token for second line: `select * from #test` and returns True
5. we can correctly fetch all rows
6. we call `nextset` again, which calls `find_result_or_done`, it finds DONE token for third line: `update #test set a = 2` and returns False, which is passed to user
7. `nextset` returned False, we stop processing even though there is one more resultset

denisenkom commented:
BTW if you want to not to call nextset for empty recondsets you can use "SET NOCOUNT ON" https://msdn.microsoft.com/en-us/library/ms189837.aspx

VFLashM commented:
The problem is not that I have to call nextset().

The problem is that neither fetch*() nor nextset() work after assigning select.

denisenkom commented:
Here is how it should look like:
```
                if self.done_flags & TDS_DONE_MORE_RESULTS:
                    if self.done_flags & TDS_DONE_COUNT:
                        return True
                    else:
                        # skip results without rowcount
                        continue
                else:
                    return False
```
Apparently there is no test which tests such situation, can you add one?

VFLashM commented:
Sure, it makes sense this way.
I changed the code and added a test for query with nocount enabled.

On a side note, process_rpc is suspiciously similar to find_result_or_done.
It seems to me that difference meant to skip resultsets from inside the procedure, but that's not the case now. Moreover, stored procedures can't really work the way dbapi expect them to work.

If I ever have enough time for that, I'd like to revise the way stored procedure call works and make sure that we're not loosing return values.

denisenkom commented:
I got test failure:

```
======================================================================
ERROR: test_stored_proc (tests.tests.DbTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/misha/pytds/tests/tests.py", line 473, in test_stored_proc
    values = cur.callproc('testproc', (val, default, output(value=1)))
  File "/Users/misha/pytds/pytds/__init__.py", line 497, in callproc
    return self._callproc(procname, parameters)
  File "/Users/misha/pytds/pytds/__init__.py", line 478, in _callproc
    parameters = self._session._convert_params(parameters)
  File "/Users/misha/pytds/pytds/tds.py", line 3201, in _convert_params
    params.append(self.make_param('', parameter))
  File "/Users/misha/pytds/pytds/tds.py", line 3186, in make_param
    column.type = self._autodetect_column_type(value, value_type)
  File "/Users/misha/pytds/pytds/tds.py", line 3144, in _autodetect_column_type
    raise DataError('Parameter type is not supported: {!r} {!r}'.format(value, value_type))
DataError: Parameter type is not supported: None <type 'instance'>
```
Is it caused by this new change?

VFLashM commented:
It's caused by previous change (my bad) and fixed in another pull request (the one that caused problems with NoneType on python3)

denisenkom commented:
Большое Спасибо!
